typescript/packages/subsurface-viewer/src/layers/grid3d/webworker.ts

```diff
@@ -839,6 +833,95 @@ export function makeFullMesh(e: { data: WebWorkerParams }) {
+    }
+
+    interface IMeshArrays {
+        trianglePoints: Float32Array;
+        triangleNormals: Float32Array;
+        properties: Float32Array;
+        lineIndices: Uint32Array;
+    }
+
+    const getPrimitiveCounts = (polys: Uint32Array): IPrimitiveCounts => {
+        let triangles = 0;
+        let lineSegments = 0;
+        let i = 0;
+        while (i < polys.length) {
+            triangles += polys[i] - 2;
+            lineSegments += polys[i];
+            i += polys[i] + 1;
+        }
+        return { triangles, lineSegments };
+    };
+
+    const tryCreateArrays = (counts: IPrimitiveCounts): IMeshArrays | null => {
+        try {
+            const trianglePoints = new Float32Array(counts.triangles * 9); // 3 points * 3 coordinates per point per 1 triangle
+            const triangleNormals = new Float32Array(counts.triangles * 9); // 3 points * 3 coordinates per point per 1 triangle
+            const properties = new Float32Array(counts.triangles * 3); // 3 points per 1 triangle
+            const lineIndices = new Uint32Array(counts.lineSegments * 2); // 2 point indices per segment
+            return {
+                trianglePoints,
+                triangleNormals,
+                properties,
+                lineIndices,
+            };
+        } catch (error) {
+            console.log(error);
+            return null;
+        }
+    };
+
+    const createMeshArrays = (
+        counts: IPrimitiveCounts
+    ): { arrays: IMeshArrays | null; counts: IPrimitiveCounts } | null => {
+        const currentCounts = {
+            ...counts,
+        };
+        let res: IMeshArrays | null = null;
+        do {
+            res = tryCreateArrays(currentCounts);
+            if (res === null) {
+                currentCounts.triangles -= counts.triangles * 0.1;
+                currentCounts.lineSegments -= counts.lineSegments * 0.1;
+                console.warn("Arrays size reduced!");
+            }
+        } while (res === null && currentCounts.triangles > 0);
+        return {
+            arrays: res,
+            counts: currentCounts,
+        };
+    };
+
+    const createEmptyMeshes = () => {
+        const mesh: MeshType = {
+            drawMode: 4, // corresponds to GL.TRIANGLES,
+            attributes: {
+                positions: { value: new Float32Array(), size: 3 },
+                properties: { value: new Float32Array(), size: 1 },
+                normals: { value: new Float32Array(), size: 3 },
+            },
+            vertexCount: 0,
+        };
+
+        const mesh_lines: MeshTypeLines = {
+            drawMode: 1, // corresponds to GL.LINES,
+            attributes: {
+                positions: { value: new Float32Array(), size: 3 },
+                indices: { value: new Uint32Array(), size: 1 },
+            },
+            vertexCount: 0,
+        };
+        return [
+            mesh,
+            mesh_lines,
+            [propertyValueRangeMin, propertyValueRangeMax],
```

Comment:
Why the range here ?

Reply:
Properties are per face in the input array.  As the webworker assignes a property value for every vertex it also can calc its range and return to the main thread

Reply:
These are never used. If you rename them, it still does work :/

Reply:
They are expected to be returned from the webworker

Reply:
OK, I miss-read the PR.